src/builtins/fish_indent.rs

```diff
@@ -0,0 +1,1200 @@
+                if [Category::branch, Category::list].contains(&node.category()) {
+                    node.accept(self, false);
+                    return;
+                }
+                panic!("unexpected node type");
+            }
+        }
+    }
+}
+
+/// Return whether a character at a given index is escaped.
+/// A character is escaped if it has an odd number of backslashes.
+fn char_is_escaped(text: &wstr, idx: usize) -> bool {
+    count_preceding_backslashes(text, idx) % 2 == 1
+}
+
+pub fn main() {
+    PROGRAM_NAME.set(L!("fish_indent")).unwrap();
+    panic_handler(throwing_main)
+}
+
+fn throwing_main() -> i32 {
+    // TODO: Duplicated with fish_key_reader
+    use crate::io::FdOutputStream;
+    use crate::io::IoChain;
+    use crate::io::OutputStream::Fd;
+    use libc::{STDERR_FILENO, STDOUT_FILENO};
+
+    topic_monitor_init();
+    threads::init();
+
+    let mut out = Fd(FdOutputStream::new(STDOUT_FILENO));
+    let mut err = Fd(FdOutputStream::new(STDERR_FILENO));
+    let io_chain = IoChain::new();
+    let mut streams = IoStreams::new(&mut out, &mut err, &io_chain);
+    // Using the user's default locale could be a problem if it doesn't use UTF-8 encoding. That's
+    // because the fish project assumes Unicode UTF-8 encoding in all of its scripts.
+    //
+    {
+        let s = CString::new("").unwrap();
+        unsafe { libc::setlocale(LC_ALL, s.as_ptr()) };
+    }
+    env_init(None, true, false);
+    let args: Vec<WString> = std::env::args_os()
+        .map(|osstr| str2wcstring(osstr.as_bytes()))
+        .collect();
+
+    // Only set these here so you can't set them via the builtin.
+    if let Some(features_var) = EnvStack::globals().get(L!("fish_features")) {
+        for s in features_var.as_list() {
+            future_feature_flags::set_from_string(s.as_utfstr());
+        }
+    }
+
+    do_indent(&mut streams, args)
+}
+
+pub fn fish_indent(_parser: &Parser, streams: &mut IoStreams, args: &mut [&wstr]) -> Option<c_int> {
+    let args = args.iter_mut().map(|x| x.to_owned()).collect();
+    Some(do_indent(streams, args))
+}
+
+fn do_indent(streams: &mut IoStreams, args: Vec<WString>) -> i32 {
+    // Types of output we support
+    #[derive(Eq, PartialEq)]
+    enum OutputType {
+        PlainText,
+        File,
+        Ansi,
+        PygmentsCsv,
+        Check,
+        Html,
+    }
+
+    let mut output_type = OutputType::PlainText;
+    let mut output_location = L!("");
+    let mut do_indent = true;
+    let mut only_indent = false;
+    let mut only_unindent = false;
+    // File path for debug output.
+    let mut debug_output = None;
+
+    let short_opts: &wstr = L!("+d:hvwicD:");
+    let long_opts: &[WOption] = &[
+        wopt(L!("debug"), ArgType::RequiredArgument, 'd'),
+        wopt(L!("debug-output"), ArgType::RequiredArgument, 'o'),
+        wopt(L!("debug-stack-frames"), ArgType::RequiredArgument, 'D'),
```

Comment:
I am inclined to remove these debug options - they've never been very useful and are awkward from inside fish. The third is also unused and has been for years now.

I would also move "--version" to the throwing_main because it is weird to have a builtin write its version?

Reply:
Note: `fish_indent --version` is called from doc_src/conf.py to determine the version for the docs.

This should probably be passed in from the build script via an environment variable, like it already is in #10367.